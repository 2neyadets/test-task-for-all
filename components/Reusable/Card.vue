<template lang="pug">
  .card.q-ma-sm(:class="'index' + index")
    .content.full-width.full-height.column
      .col-auto
        h2(:style="`font-size: ${fontSizeForTitle(note.title.length)}px;`")
          span {{note.title}}
      .col.q-py-sm.column.items-center(:class="note.tasks && note.tasks.length ? 'justify-between' : 'justify-center'")
        template(v-if="note.tasks && note.tasks.length")
          .col.full-width.row.justify-start.items-center(style="font-size: 12px;")
            template(v-for="(task, index) in note.tasks")
              .col-12.flex.justify-start
                MyCheckbox(
                  :key="task.id + index"
                  :task="task"
                  :label="task.text"
                  disable
                )
        template(v-else)
          p В заметке пока нет задач.
      .col-auto
        .flex.justify-between
          Btn(@click="openEditNotePage(note.id)" size="46" icon="edit")
          Btn(@click="openDeleteDialog" size="46" icon="delete")
</template>

<script>
import DeleteNoteConfirm from '../Dialogs/DeleteNoteDialog'
import Btn from './Btn'
import Dialog from './Dialog'
import MyCheckbox from './MyCheckbox'

export default {
  name: 'Card',
  components: { MyCheckbox, DeleteNoteConfirm, Dialog, Btn },
  props: {
    note: {
      type: Object,
      required: true
    },
    index: {
      type: Number,
      required: true
    },
  },
  data () {
    return {
      showDialog: false,
    }
  },
  computed: {},
  watch: {},
  created () {},
  mounted () {},
  updated () {},
  methods: {
    openEditNotePage (id) {
      this.$store.dispatch('notes/changeCurrentNote', null)
      this.$router.push(`/${id}`)
    },
    openDeleteDialog () {
      this.$store.dispatch('notes/changeCurrentNote', this.note)
      this.$store.dispatch('changeDialogState', 'deleteNote')
    },
    fontSizeForTitle (length) {
      if (length <= 10) {
        return 30
      } else {
        return 15
      }
    },
  },
}
</script>

<style scoped lang="stylus">
  .card
    position relative
    min-height 400px
    width 288px
    background #060c21
    display flex
    justify-content center
    align-items center
    border 1px solid black

    &:before
      content ''
      position absolute
      top -2px
      left -2px
      right -2px
      bottom -2px
      background white
      transform skew(2deg, 2deg)
      z-index -1

    &.index0:before
      background linear-gradient(315deg, #ff0057, #e64a19)

    &.index1:before
      background linear-gradient(315deg, #89ff00, #00bcd4)

    &.index2:before
      background linear-gradient(315deg, #e91e63, #5d02ff)

    &.index3:before
      background linear-gradient(315deg, #ff0000, #ffc107)

    &:after
      content ''
      position absolute
      top 0
      left 0
      width 50%
      height 100%
      background rgba(255, 255, 255, .05)
      pointer-events none

    .content
      position relative
      min-height 400px
      padding 16px
      transform translateY(0)

      h2
        position relative
        max-height 60px
        top 8px
        right 0
        margin 0
        padding 0
        color rgba(255, 255, 255, .35)
        transition .5s

      p
        margin 0
        padding 0
        color white
        font-size 16px

    &:hover

      .content

        h2
          top 0

        .my-btn
          transform translateY(0)
          opacity 1
          visibility visible
</style>
