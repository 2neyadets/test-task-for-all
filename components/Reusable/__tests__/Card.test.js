import Vuex from 'vuex'
import VueRouter from 'vue-router'
import { shallowMount, createLocalVue } from '@vue/test-utils'
import Card from '../Card'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(VueRouter)
const router = new VueRouter()

describe('Card component', () => {
  let wrapper, actions, store

  const note = {
      id: 'f17304bfb263-f99f3c6',
      title: 'Название заметки',
    },
    index = 0,
    tasks = [
      {
        done: false,
        id: 'f17332538b67-f47c711',
        text: 'Текст задачи 1',
      },
    ]

  beforeEach(() => {
    actions = {
      'notes/changeCurrentNote': jest.fn(),
      changeDialogState: jest.fn(),
    }
    store = new Vuex.Store({
      actions
    })

    wrapper = shallowMount(Card, {
      localVue,
      store,
      router,
      propsData: {
        note,
        index,
      },
    })
  })

  it('mounts without errors', done => {
    localVue.config.errorHandler = done // для отлавливания ошибок внутри nextTick, если нужно что-то выпонять внутри него
    expect.assertions(1)
    expect(wrapper).toBeTruthy()
    done()
  })

  it('have corresponding class', () => {
    expect.assertions(1)
    expect(wrapper.classes(`index${index}`)).toBe(true)
  })

  it('includes note title', () => {
    expect.assertions(1)
    expect(wrapper.find('h2').html()).toContain(note.title)
  })

  it('can change font-size of title', async () => {
    expect.assertions(2)
    expect(wrapper.find('h2').attributes().style).toBe('font-size: 15px;')
    wrapper.setProps({
      note: {
        ...note,
        title: 'Название',
      },
    })
    await localVue.nextTick()
    expect(wrapper.find('h2').attributes().style).toBe('font-size: 25px;')
  })

  it('can show "no tasks" message and have corresponding class', () => {
    expect.assertions(2)
    expect(wrapper.find('.card__inner').classes('justify-center')).toBe(true)
    expect(wrapper.find('p').html()).toContain('В заметке пока нет задач.')
  })

  it('can show tasks', async () => {
    expect.assertions(6)
    expect(wrapper.find('.card__inner').classes('justify-between')).toBe(false)
    expect(wrapper.find('.tasks').exists()).toBe(false)
    wrapper.setProps({
      note: {
        ...note,
        tasks
      },
    })
    await localVue.nextTick()
    expect(wrapper.find('.card__inner').classes('justify-between')).toBe(true)
    expect(wrapper.find('.tasks').exists()).toBe(true)
    expect(wrapper.find('.task').exists()).toBe(true)
    expect(wrapper.find('MyCheckbox-stub').exists()).toBe(true)
  })

  it('call "openEditNotePage" when click on edit btn', () => {
    expect.assertions(2)
    wrapper.findAll('MyBtn-stub').at(0).trigger('click')
    expect(actions['notes/changeCurrentNote']).toHaveBeenCalled()
    expect(router.currentRoute.fullPath).toBe(`/note/${note.id}`)
  })

  it('call "openDeleteDialog" when click on delete btn', () => {
    expect.assertions(2)
    wrapper.findAll('MyBtn-stub').at(1).trigger('click')
    expect(actions['notes/changeCurrentNote']).toHaveBeenCalled()
    expect(actions.changeDialogState).toHaveBeenCalled()
  })

  it('match snapshot', () => {
    expect.assertions(1)
    expect(wrapper.element).toMatchSnapshot()
  })
})
