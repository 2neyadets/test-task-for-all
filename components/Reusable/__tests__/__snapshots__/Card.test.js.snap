// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component match snapshot 1`] = `
<div
  class="card ma-sm index0"
>
  <div
    class="content full-width full-height column"
  >
    <div
      class="card__title col-auto"
    >
      <h2
        style="font-size: 15px;"
      >
        <span>
          Название заметки
        </span>
      </h2>
    </div>
    <div
      class="card__inner col py-sm column items-center justify-center"
    >
      <p>
        В заметке пока нет задач.
      </p>
    </div>
    <div
      class="card__actions col-auto"
    >
      <div
        class="flex justify-between"
      >
        <mybtn-stub
          icon="edit"
          label=""
          size="46"
          type="button"
        />
        <mybtn-stub
          icon="delete"
          label=""
          size="46"
          type="button"
        />
      </div>
    </div>
  </div>
</div>
`;
